src/backend/distributed/utils/node_metadata.c

```diff
@@ -478,10 +529,17 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
 		deletedNodeId = workerNode->nodeId;
 	}
 
-	DeleteNodeRow(nodeName, nodePort);
-
 	DeleteAllReferenceTablePlacementsFromNode(nodeName, nodePort);
 
+	hasAnyShardPlacements = NodeHasShardPlacements(nodeName, nodePort,
+												   onlyConsiderActivePlacements);
+	if (hasAnyShardPlacements)
+	{
+		ereport(ERROR, (errmsg("you cannot remove a node which has shard placements")));
```

Comment:
Maybe add a `DeletePlacementsFromGroup(int groupId)` function that does a scan on groupid and performs `simple_heap_delete`s.